### PR TITLE
Simplify reroute counting in InternalSnapshotsInfoServiceTests

### LINKED ISSUE #63352: InternalSnapshotsInfoServiceTests.testSnapshotShardSizes failures
**Build scan**: https://gradle-enterprise.elastic.co/s/7jv4a3qbt765e/tests/:server:test/org.elasticsearch.snapshots.InternalSnapshotsInfoServiceTests/testSnapshotShardSizes#1

**Repro line**: `./gradlew ':server:test' --tests "org.elasticsearch.snapshots.InternalSnapshotsInfoServiceTests.testSnapshotShardSizes" -Dtests.seed=BC4157C7A9E33DC7 -Dtests.security.manager=true -Dtests.locale=hr-HR -Dtests.timezone=PNT -Druntime.java=11`

**Reproduces locally?**: No

**Applicable branches**: `master`

**Failure history**: 2 failures at time of writing, [build stats search](https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-7d,mode:quick,to:now))&_a=(columns:!(_source),index:e58bf320-7efd-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:'test:testSnapshotShardSizes'),sort:!(time,desc)))

**Failure excerpt**:
```
org.mockito.exceptions.verification.TooLittleActualInvocations: 	
rerouteService.reroute(<any>, <any>, <any>);	
Wanted 16 times:	
-> at org.elasticsearch.snapshots.InternalSnapshotsInfoServiceTests.testSnapshotShardSizes(InternalSnapshotsInfoServiceTests.java:170)	
But was 15 times:	
-> at org.elasticsearch.snapshots.InternalSnapshotsInfoService$FetchingSnapshotShardSizeRunnable.doRun(InternalSnapshotsInfoService.java:240)	
	
	at __randomizedtesting.SeedInfo.seed([BC4157C7A9E33DC7:C34E8D4B2E4B22CC]:0)	
	at org.elasticsearch.snapshots.InternalSnapshotsInfoServiceTests.testSnapshotShardSizes(InternalSnapshotsInfoServiceTests.java:170)	
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)	
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)	
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)	
	at java.lang.reflect.Method.invoke(Method.java:566)	
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)	
	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)	
	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)	
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)	
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)	
	at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)	
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)	
	at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)	
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)	
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)	
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)	
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)	
	at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:824)	
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:475)	
	at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)	
	at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)	
	at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)	
	at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)	
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)	
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)	
	at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)	
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)	
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)	
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)	
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)	
	at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)	
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)	
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)	
	at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)	
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)	
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)	
	at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:831)	
	at java.lang.Thread.run(Thread.java:834)
```

### PR DESCRIPTION
**Pull request by tlrx**
A test failure occurs (reported in #63352) because a test is waiting for a given number of snapshot shard sizes to exist and immediately verifies the number of triggered reroutes - but interleaving execution can happen and at the time the number of reroutes is verified all reroutes might not have been executed yet (reroute is triggered outside the synchronize block that updates the snapshot shard sizes). In addition, usage of Mockito's _verify()_ can be trappy with argument catchers and this can be simplified.

Closes #63352

### CODE REVIEW

**Comment by tlrx** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-windows

File: server/src/test/java/org/elasticsearch/snapshots/InternalSnapshotsInfoServiceTests.java [hunk 1]
```diff
@@ -120,6 +107,13 @@ public void tearDown() throws Exception {
 
     public void testSnapshotShardSizes() throws Exception {
         final int maxConcurrentFetches = randomIntBetween(1, 10);
+
+        final AtomicInteger rerouteCount = new AtomicInteger(0);
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/snapshots/InternalSnapshotsInfoServiceTests.java:111]: I wonder, couldn't we get rid of the busy asserting completely by doing this via a latch (has the slight upside that we verify exact count a little stronger with the assertion on the latch count?)?

e.g.:

```diff
diff --git a/server/src/test/java/org/elasticsearch/snapshots/InternalSnapshotsInfoServiceTests.java b/server/src/test/java/org/elasticsearch/snapshots/InternalSnapshotsInfoServiceTests.java
index a4483bd5ce2..fdcdcdb013f 100644
--- a/server/src/test/java/org/elasticsearch/snapshots/InternalSnapshotsInfoServiceTests.java
+++ b/server/src/test/java/org/elasticsearch/snapshots/InternalSnapshotsInfoServiceTests.java
@@ -120,12 +120,20 @@ public class InternalSnapshotsInfoServiceTests extends ESTestCase {
 
     public void testSnapshotShardSizes() throws Exception {
         final int maxConcurrentFetches = randomIntBetween(1, 10);
+
+        final int numberOfShards = randomIntBetween(1, 50);
+        final CountDownLatch rerouteLatch = new CountDownLatch(numberOfShards);
+        final RerouteService rerouteService = (reason, priority, listener) -> {
+            listener.onResponse(clusterService.state());
+            assertThat(rerouteLatch.getCount(), greaterThanOrEqualTo(0L));
+            rerouteLatch.countDown();
+        };
+
         final InternalSnapshotsInfoService snapshotsInfoService =
             new InternalSnapshotsInfoService(Settings.builder()
                 .put(INTERNAL_SNAPSHOT_INFO_MAX_CONCURRENT_FETCHES_SETTING.getKey(), maxConcurrentFetches)
                 .build(), clusterService, () -> repositoriesService, () -> rerouteService);
 
-        final int numberOfShards = randomIntBetween(1, 50);
         final String indexName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
         final long[] expectedShardSizes = new long[numberOfShards];
         for (int i = 0; i < expectedShardSizes.length; i++) {
@@ -163,12 +171,10 @@ public class InternalSnapshotsInfoServiceTests extends ESTestCase {
 
         latch.countDown();
 
-        assertBusy(() -> {
-            assertThat(snapshotsInfoService.numberOfKnownSnapshotShardSizes(), equalTo(numberOfShards));
-            assertThat(snapshotsInfoService.numberOfUnknownSnapshotShardSizes(), equalTo(0));
-            assertThat(snapshotsInfoService.numberOfFailedSnapshotShardSizes(), equalTo(0));
-        });
-        verify(rerouteService, times(numberOfShards)).reroute(anyString(), any(Priority.class), any());
+        assertTrue(rerouteLatch.await(10L, TimeUnit.SECONDS));
+        assertThat(snapshotsInfoService.numberOfKnownSnapshotShardSizes(), equalTo(numberOfShards));
+        assertThat(snapshotsInfoService.numberOfUnknownSnapshotShardSizes(), equalTo(0));
+        assertThat(snapshotsInfoService.numberOfFailedSnapshotShardSizes(), equalTo(0));
         assertThat(getShardSnapshotStatusCount.get(), equalTo(numberOfShards));
 
         final SnapshotShardSizeInfo snapshotShardSizeInfo = snapshotsInfoService.snapshotShardSizes();
```

Just a suggestion :) this approach looks fine as well.

**Review by original-brownbear — APPROVED** [review]
LGTM, just one suggestion/question :)

File: server/src/test/java/org/elasticsearch/snapshots/InternalSnapshotsInfoServiceTests.java (hunk 1 above)
Comment (tlrx) [reply server/src/test/java/org/elasticsearch/snapshots/InternalSnapshotsInfoServiceTests.java]: This is a good suggestion, thanks!

**Comment by tlrx** [comment]
Thanks Armin!

**Pull request merged**